### ISSUE TITLE
`atomicapp stop` fails to stop application

### ISSUE THREAD

**Issue by surajssd** [projectatomic/atomicapp#630]
When I run atomicapp stop the application should stop!
```bash
(atomic) [vagrant@centos7-adb ~]$ ll /var/lib/atomicapp
ls: cannot access /var/lib/atomicapp: No such file or directory
(atomic) [vagrant@centos7-adb ~]$ sudo atomicapp run projectatomic/helloapache -v --provider=docker
1458041180 - [INFO] - cli/main.py - Action/Mode Selected is: run
[SNIP]
1458041181 - [DEBUG] - nulecule/main.py - Writing answers to file.
1458041181 - [DEBUG] - nulecule/main.py - FILE: /var/lib/atomicapp/projectatomic-helloapache-103ba878f1be/answers.conf.gen
1458041181 - [DEBUG] - nulecule/main.py - ANSWERS: {u'helloapache-app': {u'image': u'centos/httpd', u'hostport': 80}, 'general': {'namespace': 'default', 'provider': 'docker'}}

Your application resides in /var/lib/atomicapp/projectatomic-helloapache-103ba878f1be
Please use this directory for managing your application
```
**projectatomic-helloapache-103ba878f1be** is a directory where data resides, but then when I do `atomicapp stop` rather than stopping the existing app and removing the directory `/var/lib/atomicapp/projectatomic-helloapache-103ba878f1be` atomicapp creates a new one.

```bash
(atomic) [vagrant@centos7-adb ~]$ sudo atomicapp stop projectatomic/helloapache -v --provider=docker
1458041205 - [INFO] - cli/main.py - Action/Mode Selected is: stop
1458041205 - [DEBUG] - cli/main.py - Final parsed cmdline: stop projectatomic/helloapache -v --provider=docker
1458041205 - [DEBUG] - nulecule/main.py - NuleculeManager init app_path: /var/lib/atomicapp/projectatomic-helloapache-da5466fa95bd
1458041205 - [DEBUG] - nulecule/main.py - NuleculeManager init image: projectatomic/helloapache
1458041205 - [ERROR] - cli/main.py - Provided answers file doesn't exist: /var/lib/atomicapp/projectatomic-helloapache-da5466fa95bd/answers.conf.gen
(atomic) [vagrant@centos7-adb ~]$ ll /var/lib/atomicapp
total 0
drwxr-xr-x. 3 root root 93 Mar 15 07:26 projectatomic-helloapache-103ba878f1be
drwxr-xr-x. 2 root root  6 Mar 15 07:26 projectatomic-helloapache-da5466fa95bd
```

**Comment by surajssd** [comment]
Even `atomic cli` could not help
```bash
[vagrant@centos7-adb ~]$ sudo atomic run projectatomic/helloapache -v --provider=docker
Using default tag: latest
Trying to pull repository docker.io/projectatomic/helloapache ... latest: Pulling from projectatomic/helloapache

3690474eb5b4: Pull complete 
[SNIP]
Status: Downloaded newer image for docker.io/centos/httpd:latest

1390b2f5c5b73a0a98df4c4c90acee8b0a91e4db603c18801353bc421c9db189
2016-03-15 11:51:18,494 - [DEBUG] - main.py - Writing answers to file.
2016-03-15 11:51:18,494 - [DEBUG] - main.py - FILE: /host/var/lib/atomicapp/projectatomic-helloapache-ee0a870b0f50/answers.conf.gen
2016-03-15 11:51:18,494 - [DEBUG] - main.py - ANSWERS: {u'helloapache-app': {u'image': u'centos/httpd', u'hostport': 80}, 'general': {'namespace': 'default', 'provider': 'docker'}}

Your application resides in /var/lib/atomicapp/projectatomic-helloapache-ee0a870b0f50
Please use this directory for managing your application


[vagrant@centos7-adb ~]$ sudo atomic stop projectatomic/helloapache 
Unable to find active container 'helloapache'

[vagrant@centos7-adb ~]$ sudo docker ps -a
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS                NAMES
1390b2f5c5b7        centos/httpd        "/run-httpd.sh"     5 minutes ago       Up 5 minutes        0.0.0.0:80->80/tcp   default_centos-httpd_2551ce9f0f20
```

**Comment by cdrage** [comment]
You need to provide the path of the installed app in order to stop via `atomicapp`. This is due to the generated files being located there.

For your example it would be: `atomicapp stop var/lib/atomicapp/projectatomic-helloapache-103ba878f1be`

See https://github.com/projectatomic/atomicapp/blob/master/docs/quick_start.md for a better example :)

**Comment by surajssd** [comment]
@cdrage yes agreed but then Atomic App should not atleast create another directory.

**Comment by cdrage** [comment]
When running there needs to be a place to extract the files from the docker image and thus that's what a new directory is created.

**Comment by cdrage** [comment]
@surajssd with https://github.com/projectatomic/atomicapp/issues/642 we may be able to fix this :) (providing an id for the atomicapp instead of a dir path) if we implement label insertion for atomicapp.

**Comment by surajssd** [comment]
@cdrage yes surely.

**Comment by dustymabe** [comment]
I don't think I understand the problem here. When you stop an application you want to stop a specific deployment of the application. These different deployments are managed by their metadata files which are the directories under `/var/lib/atomicapp`.

The argument for this is that you could have many deployments of a particular type of application (i.e. 100 different wordpress applications) so stopping via a global application name doesn't make sense.

**Comment by cdrage** [comment]
If we have the metadata under their respective labels when it's initially deployed we would never have to access `/var/lib/atomicapp` in order to stop the deployment. 

Only having to provide an id or name of the deployed app.

I don't know how to do it yet, it's just a thought. But it'll help with the user experience without having to always provide the directory it's pulled from.

**Comment by surajssd** [comment]
@dustymabe yes I understand that I can have multiple deployments, but how do I know which one is which? 

And apart from that when I do `atomicapp stop` in some random location, Atomic App creates an unnecessary directory. It should rather error out saying that *no configuration found for the application* or something like that.

**Comment by dustymabe** [comment]
> @dustymabe yes I understand that I can have multiple deployments, but how do I know which one is which?

Let's use kubernetes as the platform and wordpress as the app for an example. If I want to deploy the wordpress app twice on to kubernetes I would use a separate namespace:

First deployment
- namespace: project1
- pod: wordpress

2nd deployment
- namespace: project2
- pod: wordpress

The namespace is defined in your `answers.conf` file and is also saved in the `answers.conf.gen` file when you execute `atomicapp run`. How do you know which one is which? It is a bit hard to tell which folder in the /var/lib/atomicapp/ directory corresponds to what deployment, but essentially the information is in there.

> And apart from that when I do atomicapp stop in some random location, Atomic App creates an unnecessary directory. It should rather error out saying that no configuration found for the application or something like that.

Can you give a detailed reproducer for this? If you are calling the tool correctly and it is doing this then it sounds like a bug.

**Comment by surajssd** [comment]
@dustymabe  this is how I am calling it `sudo atomicapp stop projectatomic/helloapache -v --provider=docker`, you can also see the first comment in this thread (It may be the wrong way to do it).

But it creates another directory and says it has no configuration, IMO it should either look in the current directory(where I am running the command) or there should be a mandatory option alongwith `stop`, something like `--app-location`.

So basically Atomic App is currently looking into location it has just created.

**Comment by cdrage** [comment]
@surajssd 
you have to pass a directory `sudo atomicapp stop projectatomic/helloapache -v --provider=docker /var/lib/yourapp`

**Comment by surajssd** [comment]
@cdrage that's what I am saying Atomic App should not start executing unless user provides a path to the application! Also while stopping giving name of application makes no sense then?(Correct me if I am wrong)

**Comment by dustymabe** [comment]
@cdrage I would call your example wrong. If I'm calling atomicapp directly (meaning the software lives on my system) then I can simply provide a path: `sudo atomicapp stop /path/to/yourapp/`

If the software lives in the container and I am using atomic cli then I need to provide the container name just because otherwise you don't have access to the atomic app software: `sudo atomic stop projectatomic/helloapache /path/to/yourapp/`

@surajssd does that help clear things up.. Also the commit message for 93eb316 might help.

**Comment by surajssd** [comment]
@dustymabe 
Deployed an App using Atomic CLI
```bash
[vagrant@centos7-adb ~]$ sudo atomic run projectatomic/helloapache -v
Using default tag: latest
Trying to pull repository docker.io/projectatomic/helloapache ... latest: Pulling from projectatomic/helloapache

3690474eb5b4: Pull complete 
[SNIP]
<<< end >>>

1458661293 - [DEBUG] - nulecule/main.py - Writing answers to file.
1458661293 - [DEBUG] - nulecule/main.py - FILE: /host/var/lib/atomicapp/projectatomic-helloapache-0de0ca509d38/answers.conf.gen
1458661293 - [DEBUG] - nulecule/main.py - ANSWERS: {u'helloapache-app': {u'image': u'centos/httpd', u'hostport': 80}, 'general': {'namespace': 'default', 'provider': 'kubernetes'}}

Your application resides in /var/lib/atomicapp/projectatomic-helloapache-0de0ca509d38
Please use this directory for managing your application

[vagrant@centos7-adb ~]$ ll /var/lib/atomicapp
total 0
drwxr-xr-x. 3 root root 93 Mar 22 11:41 projectatomic-helloapache-0de0ca509d38
```

Trying to stop the app
```bash
[vagrant@centos7-adb ~]$ sudo atomic stop projectatomic/helloapache  /var/lib/atomicapp/projectatomic-helloapache-0de0ca509d38/
/bin/atomic: unrecognized arguments: /var/lib/atomicapp/projectatomic-helloapache-0de0ca509d38/
Try 'atomic --help' for more information.


[vagrant@centos7-adb ~]$ sudo atomic stop --help
usage: atomic stop [-h] [-n NAME] image

positional arguments:
  image                 container image

optional arguments:
  -h, --help            show this help message and exit
  -n NAME, --name NAME  name of container

atomic will just stop the container if it is running, if image does not
specify LABEL STOP


[vagrant@centos7-adb ~]$ sudo atomic stop projectatomic/helloapache  --help
usage: atomic stop [-h] [-n NAME] image

positional arguments:
  image                 container image

optional arguments:
  -h, --help            show this help message and exit
  -n NAME, --name NAME  name of container

atomic will just stop the container if it is running, if image does not
specify LABEL STOP
```

**Comment by surajssd** [comment]
Even if I don't pass the path it works:
```bash
[vagrant@centos7-adb ~]$ sudo atomic stop projectatomic/helloapache  
Unable to find active container 'helloapache'
```

So I thought maybe it is searching for the conf in the local directory so I went into the directory where it is deployed:
```bash
[vagrant@centos7-adb ~]$ ll /var/lib/atomicapp
total 0
drwxr-xr-x. 3 root root 93 Mar 22 11:41 projectatomic-helloapache-0de0ca509d38

[vagrant@centos7-adb atomicapp]$ cd /var/lib/atomicapp/projectatomic-helloapache-0de0ca509d38/
[vagrant@centos7-adb projectatomic-helloapache-0de0ca509d38]$ sudo atomic stop projectatomic/helloapache  
Unable to find active container 'helloapache'
```
That didn't help either

```bash
[vagrant@centos7-adb atomicapp]$ docker ps
CONTAINER ID        IMAGE                                                        COMMAND             CREATED             STATUS              PORTS                NAMES
05a83c24172f        centos/httpd                                                 "/run-httpd.sh"     15 minutes ago      Up 15 minutes                            k8s_helloapache.87612f8f_helloapache_default_8e1f837a-f044-11e5-a6cd-525400061d55_0a01fa35
efc9c25fcbb1        registry.access.redhat.com/rhel7/pod-infrastructure:latest   "/pod"              16 minutes ago      Up 16 minutes       0.0.0.0:80->80/tcp   k8s_POD.859a0aaf_helloapache_default_8e1f837a-f044-11e5-a6cd-525400061d55_316a197c

[vagrant@centos7-adb projectatomic-helloapache-0de0ca509d38]$ sudo atomic stop projectatomic/helloapache  .
/bin/atomic: unrecognized arguments: .
Try 'atomic --help' for more information.
[vagrant@centos7-adb projectatomic-helloapache-0de0ca509d38]$ 
```

**Comment by surajssd** [comment]
AtomicApp works totally fine in that respect.
```bash
[vagrant@centos7-adb apache-centos7-atomicapp]$ pwd
/home/vagrant/nulecule-library/apache-centos7-atomicapp
[vagrant@centos7-adb apache-centos7-atomicapp]$ ll
total 12
drwxrwxr-x. 5 vagrant vagrant   51 Mar 22 12:42 artifacts
-rw-rw-r--. 1 vagrant vagrant  319 Mar 22 12:42 Dockerfile
-rw-rw-r--. 1 vagrant vagrant  717 Mar 22 12:42 Nulecule
-rw-rw-r--. 1 vagrant vagrant 2874 Mar 22 12:42 README.md
[vagrant@centos7-adb apache-centos7-atomicapp]$ sudo atomicapp run . -v 
1458665035 - [INFO] - cli/main.py - Action/Mode Selected is: run
1458665035 - [DEBUG] - cli/main.py - Final parsed cmdline: run . -v
1458665035 - [DEBUG] - nulecule/main.py - NuleculeManager init app_path: .
1458665035 - [DEBUG] - nulecule/main.py - NuleculeManager init image: None
[SNIP]
1458665035 - [DEBUG] - nulecule/main.py - ANSWERS: {u'apache-centos7-atomicapp': {u'image': u'centos/httpd', u'hostport': 80}, 'general': {'namespace': 'default', 'provider': 'kubernetes'}}

Your application resides in .
Please use this directory for managing your application

[vagrant@centos7-adb apache-centos7-atomicapp]$ ll
total 16
-rw-r--r--. 1 root    root     113 Mar 22 12:43 answers.conf.gen
drwxrwxr-x. 5 vagrant vagrant   51 Mar 22 12:42 artifacts
-rw-rw-r--. 1 vagrant vagrant  319 Mar 22 12:42 Dockerfile
-rw-rw-r--. 1 vagrant vagrant  717 Mar 22 12:42 Nulecule
-rw-rw-r--. 1 vagrant vagrant 2874 Mar 22 12:42 README.md
[vagrant@centos7-adb apache-centos7-atomicapp]$ sudo atomicapp stop .
[INFO] - main.py - Action/Mode Selected is: stop
[WARNING] - plugin.py - Configuration option 'providerconfig' not found
[WARNING] - plugin.py - Configuration option 'providerconfig' not found
[INFO] - kubernetes.py - Using namespace default
[INFO] - kubernetes.py - trying kubectl at /usr/bin/kubectl
[INFO] - kubernetes.py - found kubectl at /usr/bin/kubectl
[INFO] - kubernetes.py - Undeploying from Kubernetes
[vagrant@centos7-adb apache-centos7-atomicapp]$ 
```

**Comment by dustymabe** [comment]
@surajssd There was an issue with atomic cli where it didn't properly handle stop. I fixed it in this PR https://github.com/projectatomic/atomic/pull/263 . Can you try with Atomic CLI >= 1.8 and verify the behavior is what you expect?

I don't think centos has atomic cli >=1.8 yet so you may have to try on Fedora.

**Comment by surajssd** [comment]
@dustymabe yes will give it a try, thanks for the pointer.

One more final thing:

Deployed an app using AtomicApp and the container
```bash
[vagrant@centos7-adb ~]$ sudo atomicapp run projectatomic/helloapache
[INFO] - main.py - Action/Mode Selected is: run
[INFO] - base.py - Unpacking image: projectatomic/helloapache to /var/lib/atomicapp/projectatomic-helloapache-6d5770341ca0
[INFO] - container.py - Pulling Docker image: projectatomic/helloapache
Using default tag: latest
Trying to pull repository docker.io/projectatomic/helloapache ... latest: Pulling from projectatomic/helloapache

3690474eb5b4: Pull complete 
[SNIP]
[INFO] - kubernetes.py - trying kubectl at /usr/bin/kubectl
[INFO] - kubernetes.py - found kubectl at /usr/bin/kubectl
[INFO] - kubernetes.py - Deploying to Kubernetes

Your application resides in /var/lib/atomicapp/projectatomic-helloapache-6d5770341ca0
Please use this directory for managing your application
```


Here this should error out on the argparse itself, because I have not specified the path but still it is running. Make the path mandatory that is my suggestion. And magically it is looking in a directory which was never there, so this also means Atomic App created it (like just now).
```bash
[vagrant@centos7-adb ~]$ sudo atomicapp stop projectatomic/helloapache
[INFO] - main.py - Action/Mode Selected is: stop
[ERROR] - main.py - Provided answers file doesn't exist: /var/lib/atomicapp/projectatomic-helloapache-c7902949d485/answers.conf.gen
[vagrant@centos7-adb ~]$ 
```

**Comment by dustymabe** [comment]
```
sudo atomicapp run projectatomic/helloapache
```

This is doing a combination of "fetch" and "run": https://github.com/projectatomic/atomicapp/blob/master/docs/atomicapp_lifecycle.md#fetch

> Here this should error out on the argparse itself, because I have not specified the path but still it is running. Make the path mandatory that is my suggestion.

This would make command a lot longer.. The user has the option of controlling the destination with the `--destination` option. 

```
sudo atomicapp stop projectatomic/helloapache
```

> And magically it is looking in a directory which was never there, so this also means Atomic App created it

This is probably a bug.

**Comment by surajssd** [comment]
Now I have this package installed `atomic-1.8-3.gitcc5997a.fc23.x86_64` and it works like charm.

Installed:
```bash
[vagrant@fedora ~]$ sudo atomic run projectatomic/helloapache -v --provider=docker
Using default tag: latest
Trying to pull repository docker.io/projectatomic/helloapache ... latest: Pulling from projectatomic/helloapache

1544084fad81: Pull complete 
474c2ee77fa3: Pull complete 
6ee235cf4473: Pull complete 
[SNIP]
1458722528 - [DEBUG] - nulecule/main.py - Writing answers to file.
1458722528 - [DEBUG] - nulecule/main.py - FILE: /host/var/lib/atomicapp/projectatomic-helloapache-7dcf59b93594/answers.conf.gen
1458722528 - [DEBUG] - nulecule/main.py - ANSWERS: {u'helloapache-app': {u'image': u'centos/httpd', u'hostport': 80}, 'general': {'namespace': 'default', 'provider': 'docker'}}

Your application resides in /var/lib/atomicapp/projectatomic-helloapache-7dcf59b93594
Please use this directory for managing your application
```
Stopped:
```bash
[vagrant@fedora ~]$ sudo atomic stop projectatomic/helloapache -v --provider=docker /var/lib/atomicapp/projectatomic-helloapache-7dcf59b93594
docker run -it --rm  --privileged -v /home/vagrant:/atomicapp -v /run:/run -v /:/host --net=host --name helloapache -e NAME=helloapache -e IMAGE=projectatomic/helloapache projectatomic/helloapache  stop  -v --provider=docker /var/lib/atomicapp/projectatomic-helloapache-7dcf59b93594
Usage of loopback devices is strongly discouraged for production use. Either use `--storage-opt dm.thinpooldev` or use `--storage-opt dm.no_warn_on_loop_devices=true` to suppress this warning.
1458722767 - [INFO] - cli/main.py - Action/Mode Selected is: stop
1458722767 - [DEBUG] - cli/main.py - Final parsed cmdline: stop -v --provider=docker /var/lib/atomicapp/projectatomic-helloapache-7dcf59b93594
1458722767 - [DEBUG] - nulecule/main.py - NuleculeManager init app_path: /host/var/lib/atomicapp/projectatomic-helloapache-7dcf59b93594
1458722767 - [DEBUG] - nulecule/main.py - NuleculeManager init image: None
1458722767 - [DEBUG] - atomicapp/utils.py - Loading answers from file: /host/var/lib/atomicapp/projectatomic-helloapache-7dcf59b93594/answers.conf.gen
1458722767 - [DEBUG] - atomicapp/plugin.py - Loading providers from /opt/atomicapp/atomicapp/providers
1458722767 - [DEBUG] - atomicapp/plugin.py - Loading providers from /opt/atomicapp/atomicapp/providers
1458722767 - [DEBUG] - atomicapp/plugin.py - Found provider <class 'docker.DockerProvider'>
1458722767 - [WARNING] - atomicapp/plugin.py - Configuration option 'providerconfig' not found
1458722767 - [DEBUG] - atomicapp/plugin.py - Found provider <class 'docker.DockerProvider'>
1458722767 - [WARNING] - atomicapp/plugin.py - Configuration option 'providerconfig' not found
1458722767 - [DEBUG] - providers/docker.py - Given config: {u'image': u'centos/httpd', u'hostport': 80, u'namespace': u'default', u'provider': u'docker'}
1458722767 - [DEBUG] - providers/docker.py - Namespace: default
1458722767 - [INFO] - providers/docker.py - Undeploying to provider: Docker
1458722767 - [INFO] - providers/docker.py - Stopping container: default_centos-httpd_79d3718a5591
default_centos-httpd_79d3718a5591
```

**Comment by surajssd** [comment]
Closing in favor of #683 thanks @dustymabe and @cdrage for the inputs here.
